libtiledbsoma/src/utils/fastercsx.h

```diff
@@ -0,0 +1,543 @@
+
+            if (idx_end < idx_start || idx_end > nnz)
+                throw std::overflow_error("Row pointer exceeds nnz");
+
+            std::vector<std::pair<CSX_MINOR_IDX, VALUE>> temp(
+                idx_end - idx_start);
+            for (uint64_t n = 0, idx = idx_start; idx < idx_end; ++n, ++idx) {
+                temp[n] = std::make_pair(Bj[idx], Bd[idx]);
+            }
+
+            std::sort(
+                temp.begin(), temp.end(), index_lt_<CSX_MINOR_IDX, VALUE>);
+
+            for (uint64_t n = 0, idx = idx_start; idx < idx_end; ++n, ++idx) {
+                Bj[idx] = temp[n].first;
+                Bd[idx] = temp[n].second;
+            }
+
+            return Status::Ok();
+        });
+
+    assert(status.ok());
+};
+
+/**
+ * @brief Unsafe copy from compressed to dense format, as a C-order
+ * contiguous array
+ *
+ */
+template <class VALUE, class CSX_MINOR_IDX, class CSX_MAJOR_IDX>
+void copy_to_dense(
+    ThreadPool* const tp,
+    uint64_t major_start,
+    uint64_t major_end,
+    const Shape& shape,
+    Format cm_format,
+    const tcb::span<const CSX_MAJOR_IDX>& Bp,
+    const tcb::span<const CSX_MINOR_IDX>& Bj,
+    const tcb::span<const VALUE>& Bd,
+    tcb::span<VALUE> out) {
+    auto [n_row, n_col] = shape;
+    assert(Bp.size() > ((cm_format == Format::CSR) ? n_row : n_col));
+    assert(major_start <= ((cm_format == Format::CSR) ? n_row : n_col));
+    assert(major_end <= ((cm_format == Format::CSR) ? n_row : n_col));
+    if (major_start >= major_end)
+        return;
+
+    if (cm_format == Format::CSR) {
+        uint64_t out_n_col = n_col;
+        auto status = parallel_for(
+            tp,
+            major_start,
+            major_end,
+            [&Bp, &Bj, &Bd, &major_start, &out_n_col, &out](uint64_t i) {
+                uint64_t p_start = Bp[i];
+                uint64_t p_stop = Bp[i + 1];
+                uint64_t out_row = i - major_start;
+                for (uint64_t j = p_start; j < p_stop; ++j) {
+                    uint64_t out_col = Bj[j];
+                    uint64_t out_idx = out_row * out_n_col +
+                                       out_col;  // C ordered
+                    if (out_idx >= out.size()) [[unlikely]]
+                        throw std::overflow_error(
+                            "Out array is too small for provided "
+                            "coordinate "
+                            "range.");
+                    out[out_idx] = Bd[j];
+                }
+                return Status::Ok();
+            });
+        assert(status.ok());
+    } else {
+        // parallelizing this is less beneficial than the CSR case as it
+        // partitions by columns, but writes in C order. This does not align
+        // with CPU cache lines, and is therefore quite a bit slower than
+        // the speedup achieved by the (above) CSR loop.
+        //
+        // We could write a Fortran-ordered output array, but then it would
```

Comment:
I know (and you know) C order == row-major and Fortran order == col-major but it mightn't hurt to simply say row-major and col-major here in alignment with core conventions. Just a nit; just a thought.

Reply:
As you mean them, row-major and col-major are Tile concepts.  The memory data structure term of art is being used (see numpy docs for example).

However, I'll add both to hit all readers